test/unit/math/prim/fun/binary_scalar_tester.hpp

```diff
@@ -0,0 +1,148 @@
+ * framework returns the same value as the function with scalar inputs,
+ * for all valid combinations of scalar/vector/nested vector.
+ *
+ * @tparam F Type of functor to apply.
+ * @tparam T1 Type of first vector.
+ * @tparam T2 Type of second vector.
+ * @param x First vector input to which operation is applied.
+ * @param y Second vector input to which operation is applied.
+ * @param f functor to apply to inputs.
+ */
+template <typename F, typename T1, typename T2,
+          require_all_not_std_vector_t<T1, T2>* = nullptr>
+void binary_scalar_tester_impl(const F& f, const T1& x, const T2& y) {
+  auto vec_vec = f(x, y);
+  auto vec_scal = f(x, y(1));
+  auto scal_vec = f(x(1), y);
+  for (int i = 0; i < x.size(); ++i) {
+    EXPECT_FLOAT_EQ(f(x(i), y(i)), vec_vec(i));
+    EXPECT_FLOAT_EQ(f(x(i), y(1)), vec_scal(i));
+    EXPECT_FLOAT_EQ(f(x(1), y(i)), scal_vec(i));
+  }
+  std::vector<T1> nest_x{x, x, x};
+  std::vector<T2> nest_y{y, y, y};
+  auto nestvec_nestvec = f(nest_x, nest_y);
+  auto nestvec_scal = f(nest_x, y(1));
+  auto scal_nestvec = f(x(1), nest_y);
+  for (int i = 0; i < 3; ++i) {
+    for (int j = 0; j < x.size(); ++j) {
+      EXPECT_FLOAT_EQ(f(nest_x[i](j), nest_y[i](j)), nestvec_nestvec[i](j));
+      EXPECT_FLOAT_EQ(f(nest_x[i](j), y(1)), nestvec_scal[i](j));
+      EXPECT_FLOAT_EQ(f(x(1), nest_y[i](j)), scal_nestvec[i](j));
+    }
+  }
+  std::vector<std::vector<T1>> nest_nest_x{nest_x, nest_x, nest_x};
+  std::vector<std::vector<T2>> nest_nest_y{nest_y, nest_y, nest_y};
+  auto nestnestvec_nestnestvec = f(nest_nest_x, nest_nest_y);
+  auto nestnestvec_scal = f(nest_nest_x, y(1));
+  auto scal_nestnestvec = f(x(1), nest_nest_y);
+  for (int i = 0; i < 3; ++i) {
+    for (int j = 0; j < 3; ++j) {
+      for (int k = 0; k < x.size(); ++k) {
+        EXPECT_FLOAT_EQ(f(nest_nest_x[i][j](k), nest_nest_y[i][j](k)),
+                        nestnestvec_nestnestvec[i][j](k));
+        EXPECT_FLOAT_EQ(f(nest_nest_x[i][j](k), y(1)),
+                        nestnestvec_scal[i][j](k));
+        EXPECT_FLOAT_EQ(f(x(1), nest_nest_y[i][j](k)),
+                        scal_nestnestvec[i][j](k));
+      }
+    }
+  }
+}
+
+/**
+ * Implementation function which checks that the binary vectorisation
+ * framework returns the same value as the function with scalar inputs,
+ * for all valid combinations of scalar/vector/nested vector.
+ *
+ * This is a specialisation for std::vector inputs.
+ *
+ * @tparam F Type of functor to apply.
+ * @tparam T1 Type of first vector.
+ * @tparam T2 Type of second vector.
+ * @param x First vector input to which operation is applied.
+ * @param y Second vector input to which operation is applied.
+ * @param f functor to apply to inputs.
+ */
+template <typename F, typename T1, typename T2,
+          require_all_std_vector_t<T1, T2>* = nullptr>
+void binary_scalar_tester_impl(const F& f, const T1& x, const T2& y) {
+  auto vec_vec = f(x, y);
+  auto vec_scal = f(x, y[1]);
```

Comment:
Why dou you use the second element for the scalar? Would it not make more sence tu use the first one, so this can also be called with a vector of length 1?

Reply:
This was just a side-effect of spending too much time in R for work, now fixed.